added Center Mode (#15684)

SrTobi opened the pull request:
As described in #15684 this adds editor centering for the zen mode. I added two options:
- `zenMode.centerEditor` controls if the centering should be used
- `zenMode.centerEditorSize` controls the size of the centered editor. you can either specify the size using a size string (e.g. `"60%"` or `"600px"`) or you can specify the size of the two halfs of the editor (seen from the center)

```
"zenMode.centerEditorSize": {
  "left": "30%",
  "right": "100%"
}
```
With this the editor would occupy 30% of the left side of the window and 100% of the right side.

![preview](https://github.com/SrTobi-Forks/vscode/blob/SrTobi/center-zen-preview/resources/center-zen-preview.gif?raw=true)

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=40757) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=40757) before we can accept your contribution.<br/><br/>:x: SrTobi [sign now](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=40757)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vscode?pullRequest=40757) it.</sub>

SrTobi commented:
It says that `editorGroupsControl.ts is not formatted`. But I don't know why (and it doesn't say why)... I had to commit it via `git commit --no-verify` because I couldn't figure this out... If I use the buildin vscode `Format Code` command it destroys the whole file... is there some script that does the formating?

~~~~
[15:51:40] Using gulpfile ~/build/Microsoft/vscode/gulpfile.js
[15:51:40] Starting 'hygiene'...
.............................................................................................................../home/travis/build/Microsoft/vscode/src/vs/workbench/browser/parts/editor/editorGroupsControl.ts is not formatted
.....................................................
[15:53:27] 'hygiene' errored after 1.77 min
[15:53:27] Error: Hygiene failed with 1 errors. Check 'build/gulpfile.hygiene.js'.
~~~~

isidorn commented:
@SrTobi thanks a lot for your PR. This functionality is needed and I really appreciated that you submited a PR. However it is usually best to first provide an empty PR where we can discuss on how to potentially solve this - just so you know for the future.

For now I did a quick review, and the issue that I see is that the layouting magic is happenig in the `editorGroupsControl.ts` I believe all the layouting math should be done in the [`WorkbenchLayout`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/browser/layout.ts#L50) and the workbench layout should just pass the apropraite dimension to the editor part.
However I also leave this up to @bpasero to comment since he owns the `EditorsGroupControl`

As for the settings and options they are fine overall, I would just polish them a bit, but let's first get the layouting correct.

SrTobi commented:
@isidorn I was aware that the layouting happens in editorGroupsControl but I specifically wanted the tab bar taking the full space and not being centered as well (which would be the case if I had centered the whole editorGroupsControl). That's why I had to move it into editorGroupsControl. That was simply a design decision and could be done either way.

This becomes even more important with the new idea to also make the centered mode available in tho non-zen-mode (which we discussed in #15684). The settings we worked out there are more fitting for the use cases of most users, I think.

bpasero commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -86,6 +89,23 @@ export interface IEditorGroupsControl {
 	dispose(): void;
 }
 
+function toSize(sizeStr: string, maxSize: number, defaultSize: number): number {
```
Any reason why this method is not a private method within the `EditorGroupsControl` close to where it is being used?

bpasero commented:
I think the approach is fine to have it inside `EditorGroupsControl` and it seems to work OK. However what is not clear to me and needs to be decided is a) under which condition this setting should apply and b) how the setting looks like. 

a) from what other people seem to say they want to center the editor independently from zen mode and I think it is fine to allow this independent from it

b) some seem to want a setting of the size based on character count and not relative percentages.  I also wonder if maybe as a user you should be able to resize the editor using sashes to the left and right to configure the right size without having to manually edit this in settings

mjbvz commented on src/vs/workbench/electron-browser/main.contribution.ts:
```diff
@@ -418,6 +418,28 @@ configurationRegistry.registerConfiguration({
+			'default': false,
+			'description': nls.localize('zenMode.centerEditor', "Controls if a single editor is centered when in zen mode.")
+		},
+		'zenMode.centerEditorSize': {
```
I'm in favor of a simpler approach where the max centered editor size is based on columns,  as with `editor.wordWrapColumn`

SrTobi commented:
@bpasero I like the idea with the sashes, but then the question when to activate the center mode becomes even harder, or not? Maybe combine both ideas:

1. The user defines a number `N` of characters
2. We calculate the width `W` of an EditorGroupsControl which holds exactly `N` characters in one line
3. If the available space is greater than `W` => activate the centered mode
4. Dependent on the user settings do not limit the editor size on the right side
5. Instead of using fixed borders add sashes left and right (right only if the user has deactivated 4.)

Then two questions remain:
- Do we save the width adjusted by the user with the sashes, or not. And if we save it, does it have any effect on our decision when to activate the centered mode?
- What do we do with multiple EditorGroups? Do we simply deactivate the centered mode (as the current PR does it) or do we center both of them?

And then one technical question:
- How to calculate `W` in step 2. Does a function exists that does that? I mean it's not done with `N * fontWidth` because there might be line numbers and a scroll bar and stuff. We could, of course, ignore that when calculating `W`.

SrTobi commented:
Ok, I changed the PR and streamlined the settings a lot.

The Center Mode is now independent from the zen mode.
It can be activated automatically or manually by the "Toggle Center Mode" action.
If active, it will set the width of the active silo to `centerMode.size` and center it.

The settings are now as follows:

- `centerMode.autoActivate` has the options `always`,  `zen`, `fullscreen` and `never`. It controls in which context the center mode should be activated automatically.
- `centerMode.size` controls the size of the centered silo.
- `centerMode.adoptRight' -> Normally, on the right side of the centered silo would be free space. With this option the silo can be expanded all the way to the right border
- `centerMode.onlyEditors` -> I think most people would use the centered mode only for the text editor... it doesn't make much sense - in my opinion - to use it for a diff  widget or a settings page or welcome page. But if you want that, you can set `onlyEditors` to `false`.

bpasero commented:
@SrTobi I would try to reduce the settings to begin with and make this feature simple. For that I think its good that it is decoupled from zen mode now. 

Settings to drop or question:
* `centerMode.adoptRight` drop until there is demand for it
* `centerMode.onlyEditors` drop and only enable for text editors
* `centerMode.autoActivate` this one is confusing to me, what does `always` or `never` really mean? If I compare this with zen mode, there we just have one setting: `zenMode.fullScreen` and I can understand that.

isidorn commented:
@SrTobi great work.
Mostly agree with @bpasero 
Too much settings to start. It is good that it is decoupled from zenMode
* Definetely remove `centerMode.adoptRight`, `centerMode.onlyEditors`
* I would also remove `centerMode.autoActivate` and let's start with this being a command. Also the zen mode by default will enter this mode. On top of that zenMode can have a setting if it should transition to this mode or not.
* Sorry if I missed this discussion but why are we not having 2 sashes so the user can control the horizontal size instead of the `centerMode.size` settings
* Name `centerMode` sounds a bit ugly since I would not call this a mode, sounds like it could clash with `zenMode` which is not correct
* Does the centerMode stay active if the user opens 2 editors side by side. IMHO we should consider if the centering should take place only if one editor is open and we automatically escape it if 2 or more editor groups exist. I think this is worth considering as it might behave nice for the most common use case. User has one editor, centers the thing. As soon as he does side by side we exit the center mode, but we return to it if the user closes the editors on the side.

SrTobi commented:
@bpasero The idea behind `autoActivate` was that the user might want to always enter centered mode if he uses fullscreen and/or zenmode but not when in the normal view. That's what I need the centered mode for 😄 And that's why the first version of the PR only activates it in zen mode. `never` and `always` mean exactly that: The centered mode is `always` (independent of fullscreen and zen mode) forced by the settings or `never` respectively. Independent of the settings the centered mode can than be activated by the user using the "Toggle center mode" action.

@isidorn The moment a second silo is opened, the centering is prevented. So no centering, but no settings or internal state changes either. That means, if one of the two silos is closed, the remaining silo will be centered again.

---- 

- To `centerMode.onlyEditors`: yeah that can be removed.

- To `centerMode.autoActivate`: I think we can scrap that, but then we should add an option to zen mode which activates the centered mode

- To the naming: I like the idea renaming it from centered mode to centered layout.

- To `adoptRight`: I think this is a really cool feature. If I center my editor, I want to do it because I want to look at the center of my screen. That is achieved by a left margin because at least 90% of the code are in the first 90-120 columns. With a margin on the right side, the remaining 10% of the code are now either wrapped to the next line or hidden accompanied by a horizontal scroll bar. And that despite of the fact that there is a big free space on the right side of my editor. That's why I think having only a left margin would be quite cool.

- To `adoptRight`/`size`: @bpasero The interval thing was what the first version of this PR did and I think many people criticized it for beeing too complicated and I think so now as well (even I can not easily see what you mean by `[1200, 0]` 😆).
I still like the idea with the sashes which would also mean that we can really scrap all settings for the centered layout, except the auto center in zen mode (which would of course wander into the zen mode settings) and it is more user friendly as well.

When we use sashes, then there are two more questions:
1. What do we do, when the user moves the sashes all the way to the border? Especially on the left side where the sidebar is. I don't think it's a good idea to have to sashes next to each other without space in between.
We could either prevent that and enforce a minimal margin of e.g. 8px or we deactivate the centered layout the moment the user drops the sash at the border (then we should also inform the user that the centered layout has been deactivated).

2. Do the two sashes work independently or are they coupled (meaning that when the left sash is moved to the left, the right one moves automatically to the right in order to keep the editor centered). The first one is clearly more flexible but also kind of destroys the idea of centering the editor. The user would now have to do the centering himself which is very hard to nearly impossible to get it right pixel-precise.

What do you think? Otherwise I would start implementing the following changes sometime in the next days:
- Remove all the settings and add the one for the zenmode
- Add sashes that are coupled and deactivate the center layout if they are moved all the way to the border.

isidorn commented:
@SrTobi thanks for your comments.
* yes I can add an option for zenMode to automatically go into this. But let's try to keep this PR minimal and add things on top of it later on.
* Let's work on the naming once we have done the functionality. But yeah getting rid of "mode" sounds good
* `adoptRight` might be a cool feature, but as already mentioned let's keep this PR minimal and add things on top. Once we have the basic funcionality we can debate more easily about what is still missing in seperate PRs. So I would prefer if we do not do it yet
* Now I was pondering about the coupling ant to me it seems like this is really tied with the auto hide behavior. After some pondering I think we can only auto hide if they are coupled. The downside of them being coupled is that if the user has the sidebar opened and wants his editor to be in the exact center of the monitor he can not really do that. So for me it seems like we have two options:

1. Couple sashes, auto hide on 95% (like panel height min)
2. Indipendent sashes, limit theirs size to 95% (like panel height max)

Since we already auto exit this mode when more than 1 editor is open I would vote for option 2. To give more flexibility to those huge montiors and to let the users have their editor really centerd when explorer is open.

Feature which we will add on top when the users complain about pixel perfection. Double click on sash brings them both into default position (we have this in other pars of our ux), double click again takes 2/3 of the width, double click again 3/4 double click again the default.

Let's see what @bpasero thinks before you start the implementation.
Bottom line I do not think we need any settings in this PR. Which makes naming easier.

SrTobi commented:
Yeah that sounds good, though I would also add double clicking on a sash and then dragging brings it into the coupled mode... so we can have all the features :D and it would be somewhat natural (may I add that that is exactly what my usecase would be). And don't tell me about hidden features... I didn't knew about the double clicking until you mentioned it 😄 (good to know by the way 👍)

SrTobi commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -2027,10 +2037,34 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
+
+			if (tryCentering) {
+				const size = Math.max(100, this.configurationService.getValue<number>('centerMode.size'));
+				const adoptRight = this.configurationService.getValue<string>('centerMode.adoptRight');
```
This line should be getValue<boolean>

SrTobi commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -2027,10 +2037,34 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
+				const centerOnlyEditors = this.configurationService.getValue<string>('centerMode.onlyEditors');
+				const availableWidth = this.dimension.width;
+
+				const isTextEditor = editor instanceof TextResourceEditor;
```
`instanceof` is not a good design... in fact it doesn't even work correct. Markdown editors and normal text editors seem to be of class type `TextFileEditor` (which I was not able to import because linter import rules).
Best option is to add another abstract method like `allowsCentering(): boolean` or something to `BaseEditor` and let the implementation decide whether it itself should be centered or not

bpasero commented:
I like the idea of replacing the now complicated `centerMode.autoActivate` with something like `zenMode.centerEditor` because the zen related settings already go into that direction where you can tweak these things. 

Avoiding the remaining settings and having 2 sashes left and right to the editor seems like a good thing to play around with: not sure about the coupling to be honest (we do not have that anywhere else in our UI). Maybe we could say that Ctrl+click moves both or double click does something special but imho we can also look at that later. We could also introduce 2 more sashes top and bottom of the editor so that people could size the editor area anyway they want (some people seem to want a margin to the top of the editor).

Bottom line is that we should have some good defaults out of the box that make the "centered editor" mode useful without configuration. I would probably expect that default to work in any resolution (e.g. not hardcoded to a specific number but relative to the window size). E.g. with the current 1200 value I was not even seeing the editor centering when I was working on my normal laptop screen and thought the feature was broken. 

Last, we need to figure out where to store the user configured sash positions. I guess it would make sense to store this via storage service in the GLOBAL scope and not workspace so that the user can tweak it once and have it in all workspaces?

michaeljcalkins commented:
@bpasero @SrTobi Just throwing this out there, but Sublime Text uses the longest line as the width to auto center the code so it expands as the length gets longer or shrink as it gets shorter.  Thus working no matter what file you're looking at.

SrTobi commented:
@bpasero yeah, I think GLOBAL scope is fine.

@bpasero you're right: defaults are important. If the user activates the centered layout, he must be able to see the effect immediately.
A width depending on the window size seams reasonable. I really liked the idea of @michaeljcalkins, so I tested it with subl3 and some vscode .ts files (the feature had to be activated first because line wrapping was activated and sublime kept the editor size at 80 chars) and I found it not quite satisfying. As I mentioned above long lines are not a good indicator for where the bulk of code is - column wise. One would have to calculate some "mass of text" value and align according to this value. And though I find that's very interesting, it's out of scope for this PR and I think it conflicts somewhat with the sashes idea.

Another problem with the sashes: are they proportional or absolute? That's very important when changing the window size. I think most reasonable would be to make the size of the editor (between the two sashes) absolute — so the editor size doesn't change on resize — and let the left/right margin grow/shrink proportional. That would also provide a smooth transition to the non centered layout when shrinking the window too much. In the store we would have to save the size of the editor in pixels and the proportion of the right/left margin.

On the other Hand, it does create problems with any default size which depends on the window size because different windows sizes — when the centered layout is initially activated — would yield different permanently saved editor sizes. Though I think the feature would mostly be used in zen mode and so I think it's somewhat reasonable to ignore the former mentioned problem and just set it to 50% of the window size (after zen mode is activated). After all the most important point in the default size is that the user recognizes the feature and is able to modify the size with the sashes himself.

isidorn commented:
Seems like @bpasero and me are in agreement and you have something to start with. Top and bottom sashes are step2, so not for now.
As for where to store the setting, storageScope.GLOBAL seems like a good place, since you configure this per your Monitor, not per your workspace.

@SrTobi as for the propritaonal vs absolute, the sashes are absolute and they will not change when the window resizes. And that is fine for this PR. 
After we merge in the PR we can add aditional events to listen on window resize and react appropriatly if we want to improve this.
Apart from that, you choose what you want to store in the storage. So you can also store a proportion, so when getting data from storage you restore the sashes propritonaly and not absolutely.

Bottom line, let's try to get the basic funcionality in and not worry about fancy features on top. We will nicely polish this once we have the inital pr in.

Thanks a lot

bpasero commented:
I would also tolerate a first PR where we leave the sashes out for the future to start simple. The editor should just adjust its size based on the outer container size dynamically. The larger the outer bounds are, the greater the margins will be.

bpasero commented:
@SrTobi another idea would be to simply let the editor contents decide how large the editor becomes. This would have the advantage that it is up to the existing settings (`editor.wordWrap`, `editor.wordWrapColumn`) to decide how large the editor will be. If the file is wrapped after `80` columns, the editor would be centred around this size (both from the left and from the right). If the user types and the line becomes longer, the editor would also grow. 

Since e.g. markdown is configured to wrap by default, I think this could work nicely.

SrTobi commented:
@isidorn @bpasero ok, I had some time to tinker around and I implemented the sashes, but there are still some things left to do... I think, though, it is important that you check out the current version and look if that is acceptable.

The things I implemented:
- Two sashes to control the size of the centered editor
- Renamed the feature to "Toggle Centered Layout"
- Double click resets the sashes to the default (editor gets centered with a size of 50% of the available space)

Things that are missing:
- Save the position and the size of the centered layout to GLOBAL
- automatic entering of the centered layout when toggling zen mode
- ❗️ The sashes are not visible, yet (the only thing that hints at them is the changing courser)... I kind of like it for the zen mode actually, but it's not very user friendly
- Correctly deactivate the centered layout for non-text-editor editors... it's currently done with an `instanceof` which is kinda ugly... and I saw it even not working correctly a week ago... now it seems to work but I will have to investigate that further. 

@bpasero I think that is a nice idea for the default width... though I have no idea how to calculate the width of an editor that holds `n` characters... is there a function for that somewhere?

isidorn commented:
@SrTobi that sounds great. From the feature perspective what you implemented is a perfect first step and let's focus on that to get this PR merged.

I recommend that we tackle the missing features in seperate PRs and with those I can help you (just tell what you do not want to look into and I can implement).

As for the non visible sashes - isn't the left and right margin in a different color so they should be pretty easy to spot (similar as we have it now between editor and sidebar).
Deactivating the centered layout for non-text editors feels like it should not be something we need for first. I would remove that and only add it once we figure out it is needed.

I will try out your PR in a couple of hours and provide feedback directly in the code. @bpasero will review in a couple of days.

isidorn commented on src/vs/code/electron-main/menus.ts:
```diff
@@ -688,6 +688,7 @@ export class CodeMenu {
 
 		const fullscreen = new MenuItem(this.withKeybinding('workbench.action.toggleFullScreen', { label: this.mnemonicLabel(nls.localize({ key: 'miToggleFullScreen', comment: ['&& denotes a mnemonic'] }, "Toggle &&Full Screen")), click: () => this.windowsMainService.getLastActiveWindow().toggleFullScreen(), enabled: this.windowsMainService.getWindowCount() > 0 }));
 		const toggleZenMode = this.createMenuItem(nls.localize('miToggleZenMode', "Toggle Zen Mode"), 'workbench.action.toggleZenMode');
+		const toggleCenteredLayout = this.createMenuItem(nls.localize('miToggleCenteredLayout', "Toggle centered Layout"), 'workbench.action.toggleCenteredLayout');
```
Use upper case to be consistent with other menu entries, so "Toggle Centered Layout"

SrTobi replied:
done.

isidorn commented on src/vs/workbench/browser/actions/toggleCenteredLayout.ts:
```diff
@@ -0,0 +1,36 @@
+}
+
+const registry = Registry.as<IWorkbenchActionRegistry>(Extensions.WorkbenchActions);
+registry.registerWorkbenchAction(new SyncActionDescriptor(ToggleCenteredLayout, ToggleCenteredLayout.ID, ToggleCenteredLayout.LABEL, { primary: KeyChord(KeyMod.CtrlCmd | KeyCode.KEY_K, KeyCode.KEY_T) }), 'View: Toggle Centered Layout', nls.localize('view', "View"));
```
I would not use a default shortcut for this. So just leave undefined.
The issue with this one is that it is very similar to the CMD + K, CMD + T, which is taken by show theme
We can easily add a default keybinding later if people ask

SrTobi replied:
done.

isidorn commented on src/vs/workbench/services/part/common/partService.ts:
```diff
@@ -129,6 +129,16 @@ export interface IPartService {
+	/**
+	 * Returns whether the centered layout is active.
+	 */
+	isCenteredLayoutActive(): boolean;
```
I would rename this to `isLayoutCentered`
Notice how we do not use "active" for similar methods in the same service

SrTobi replied:
done.

isidorn commented on src/vs/workbench/electron-browser/workbench.ts:
```diff
@@ -1313,6 +1323,22 @@ export class Workbench implements IPartService {
+	public toggleCenteredLayout(): void {
+		this.centeredLayoutActive = !this.centeredLayoutActive;
+
+		if (this.centeredLayoutActive) {
```
I would replace this if statement with
`this.storageService.store(Workbench.centeredLayoutActiveStorageKey, this.centeredLayoutActive, StorageScope.GLOBAL);`

Seems more straighforwared to me.
It seems like we were doing this clunky business in zen mode so you took it from there.

SrTobi replied:
done.

SrTobi replied:
Actually you doing this 'clunky' business all the time for stored boolean values.

isidorn commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -37,6 +37,8 @@ import { attachProgressBarStyler } from 'vs/platform/theme/common/styler';
 import { IDisposable, combinedDisposable } from 'vs/base/common/lifecycle';
```
This is the only file I did not yet review. Will first try out the feature a bit.

isidorn commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -2013,6 +2129,32 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
+			this.layoutVertically &&
+			this.visibleSilos === 1 &&
+			this.partService.isCenteredLayoutActive() &&
+			this.visibleEditors[Position.ONE] instanceof TextResourceEditor;
```
This `instanceof` check is definetely not needed, it makes the behavior of this feature unpredictable, and I am not sure what is the reason for this limiation

isidorn commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -193,6 +214,10 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
 		return this.layoutVertically ? EditorGroupsControl.MIN_EDITOR_WIDTH : EditorGroupsControl.MIN_EDITOR_HEIGHT;
 	}
 
+	private get visibleSilos(): number {
```
This is not needed, use `this.editorGroupService.getStacksModel().groups.length`

SrTobi replied:
done.

isidorn commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -1858,12 +1916,70 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
+		const size = Math.min(maxSize, this.centeredEditorDragStartSize + diffSize);
+		this.setCenteredEditorPositionAndSize(size < this.minSize ? pos + (size - this.minSize) : pos, size);
+	}
+
```
This seems funky, that on `sashLeftDragEnd` you are layouting containers, and on sahRightDragEnd you do no action.
Why is layout needed when the sash event ends, it feels to me like everything is getting nicely layouting continously and an explicit call at the end is not needed

SrTobi replied:
done!? I removed the call to `this.layoutContainers` from `onCenterSashLeftDragEnd`.

isidorn commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -2043,7 +2193,16 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
+
+	private resetCenteredEditor(layout: boolean = true) {
+		this.centeredEditorLeftMarginRatio = 0.5;
+		this.centeredEditorPreferedSize = Math.floor(this.dimension.width * 0.5);
```
Let's use the [golden ratio](https://en.wikipedia.org/wiki/Golden_ratio) as the default value.
So `this.dimension.width * 0.61`

SrTobi replied:
done. nice idea.

isidorn commented:
@SrTobi great work!
The only file I did not thoroughly review is`editorGroupsControl`. I will review it in great detail tomorrow, possible after @bpasero gives it a spin.
I have tried out the feature and it behaves quite nice.

Only two feautres are needed to make this working nicely:
* You need to store the proporational position of the sashes in the storage service global, so that the sash sizes are preserved across reloades. Storing the proporation is important, not the absolute number so this works across monitors
* We need to draw a line for the sashes, or some alternative to give them more visual strength

I have also discussed with @bpasero and figuring out the inital size of the centered editor based on editor contents we will not do for now. So let's focuss on the two things needed and on the comments in the review.

Thanks again for this awesome work

SrTobi commented:
Ok, I fixed your suggestions and added the storing of the position and the location.

Things left to do:
- The `instanceOf` suggestion
- drawing the sashes. I don't know what to do here... Maybe simply add a border right and left via css?

bpasero commented:
@SrTobi there seem to be warnings in the toggle action. And then, when I try this out and select the center mode from the menu, the editor does not change, is that a bug?

isidorn commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -87,11 +90,18 @@ export interface IEditorGroupsControl {
  */
 export class EditorGroupsControl extends Themable implements IEditorGroupsControl, IVerticalSashLayoutProvider, IHorizontalSashLayoutProvider {
 
+	private static readonly CENTERED_LAYOUT_DATA_STORAGE_KEY = 'centeredLayoutData';
```
I would prefix the value of this key with 'workbench' to be aligned with similar storage keys

SrTobi replied:
done.

isidorn commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -969,6 +1015,31 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
 		// Silo Three
 		this.silos[Position.THREE] = $(this.parent).div({ class: 'one-editor-silo editor-three' });
 
+		// Center Layout stuff
```
Minor. Feels like code duplication, we do the creation and listeners twice. Can't we have a helper method that just takes a centered sash and nicely initilizes it. For that the DragStart and Drag methods would have to take sashes as parameters and based on Sash do different things.
Which might not be too bad, currently I feel like there are too many methods.

SrTobi replied:
I don't think that's a good idea... that would make the individual method body larger... and break the coding style for sashes... I mean the class is pretty big but that would require an additional PR that really breaks the class apart. I think it's much more clean the way it is... what we could do is outfactor the initialization for the sashes. I did that in this commit https://github.com/SrTobi-Forks/vscode/commit/abccb7738d4803aa4241ec9b5dab10162cbd5602 (it's not part of this PR at the moment) but I am not really sure if that makes the code more structured (in fact, it adds one more method :smile: ).

isidorn replied:
Ok, makes sense

isidorn commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -1858,12 +1929,71 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
+		this.setCenteredEditorPositionAndSize(size < this.minSize ? pos + (size - this.minSize) : pos, size);
+	}
+
+	private saveCenterdLayoutData(): void {
```
I would probably call this `storeCenteredLayoutData`

SrTobi replied:
done.

isidorn commented on src/vs/workbench/browser/parts/editor/editorGroupsControl.ts:
```diff
@@ -2013,6 +2143,32 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
+			this.layoutVertically &&
+			this.editorGroupService.getStacksModel().groups.length === 1 &&
+			this.partService.isLayoutCentered() &&
+			this.visibleEditors[Position.ONE] instanceof TextResourceEditor;
```
Yeah definetely remove this instanceof check, since the feature does not behave like it should with it. @bpasero was unable to try it out.

SrTobi replied:
done.

isidorn commented:
@SrTobi great, I added a bit more comments. @bpasero will also review that file later today I believe.
As for the borders, you can experiment with the border if you like it we can take it. An alternative is to use a different color of the left and right margin of the editor (some viewlet color perhaps).
Also you do not have to worry about this, I can also take care of it after we merge the PR.

Thanks a lot for your great work!

SrTobi commented:
@isidorn I fixed your remarks. I still think we should have something that prevents the diff editor from being centered... it really needs all the space it can get

I played a little bit around with the border for the sashes but I'm not sure on how to do it correctly... I think I would prefer a thin one pixel border with the same color the scrollbar border has (`#383838` for the normal dark+ theme).

Here is a screenshot:

![screenshot from 2018-02-20 12-15-52](https://user-images.githubusercontent.com/4148534/36421225-fc7b7ff8-1637-11e8-89f8-a9d3d71c6f24.png)

I added a commit for this but it seems somewhat messy (on the other hand: it is css :smile: ). Maybe you could enhance that further. With my solution, there also seems to be a gap between the actual scrollbar (not only the container) and the right border (it's not visible on the screenshot because the scrollbar is hidden).

isidorn commented:
@SrTobi great for addressing everything. That last commit is a total mess so I suggest to revert it 😊
The issue here is that we also have to take the theming story into account.
However the screenshot shows that this solution is not that bad. I can take care of this, you can just revert the last commit.

So when that is reverted all is good from my side, let's just wait for @bpasero before merging

As for the DiffEditor, same as when 2 groups get opened we should get out of CeneterdMode temporarly while a DiffEditor is open, however we have a [helper function](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/browser/editorBrowser.ts#L818) to check if something is a DiffEditor. You would also have to react on editors being opened if you are not already. If you feel like it give this a try, if not I can also cover it once we merge.

SrTobi commented:
@isidorn ok, I reverted the commit :D
I am not sure what you mean by "react on editors being opened" or at least I'm not sure what the problem should be. I think it's best if you take over from here. I tried to use the isDiffEditor but wasn't able to find the correct object to give as argument. I think it's easier for you, because you know the internal editor structure better than I do :smile:

When this is merged I would maybe work on the auto activation of the centered layout in zen mode. That's a feature that I would like to see very much.

isidorn commented:
@SrTobi sounds great. Let's wait for the merge and once that is done you can look into zen mode setting which should not be complicated.

SrTobi commented:
@isidorn ok, let's do it that way

bpasero commented:
LGTM, I moved the sashes into the editor container instead of the parent because I think they belong there.

isidorn commented:
@bpasero awesome, thanks for merging this in.
@SrTobi do you want to create a new PR adding setting `zenMode.centeredLayout` and I take care of the other polish?

Code pointers
https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/electron-browser/main.contribution.ts#L425
https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/electron-browser/workbench.ts#L118

SrTobi commented:
@bpasero good idea! looks great
@isidorn ok. I'll open the PR later today.

isidorn commented:
@SrTobi I have pushed the border color and not to go into center mode for diff editors.
Once we have a `zenMode` setting we will be in a good spot.
I will present this today at our meeting to get feedback, and next week we have endgame testing so I will create a test plan item to cover this feature so we test it well.

Apart from zenMode setting I would not add any other features on top until we get some user feedback.

isidorn commented:
@SrTobi the presentation in the meeting went great. Only one suggestion came: by default the sashes should be paired, but the user can press `alt` to unpair them whilce clicking and only control one sash. I will implement this now, since I think it is a great idea.

SrTobi commented:
@isidorn good idea. I really like that :+1:

dodheim commented:
Of the three languages I use VSCode for, only one of them enters center mode properly. I want to open an issue, but I'm not sure where – is this a VSCode problem, or a problem with respective extensions the two languages that aren't working for me? (F# is working, C++ and Rust aren't.)

I know this question is very light on details, but I'm not sure what information would be helpful, or whether I'm posting in the right place...

isidorn commented:
@dodheim open a new issue with exact reproducable steps and we can investigate further. Thanks
